[Feature] Add composite typography tokens

narin opened the pull request:
<!-- PR title naming convention:
'[Issue type] Brief summary of issue suitable for changelog or copy/paste issue title',
where Issue type = bug, feature, spike, CU (code upkeep), etc.-->

<!-- Preferred branch naming convention:
'[Issue type]/[Issue #]-[Your name]-[Summary of issue]',
where Issue type = bug, feature, spike, CU (code upkeep), etc.-->

## Description of Change
<!-- Describe the change and context with which it was made beyond ACs unless straightforward.
Consider:
 - What is relevant to code reviewer(s) and not in the ticket?
 - What context may be relevant to a future dev or you in 6 months about this PR?
 - Did the course of work lead to notable dead ends? If so, why didn't they pan out?
 - Did the change add new dependencies? Why?
 - Were there important sources to link? Examples: an open bug with a dependency project, an article of someone else solving the same problem that was partially or wholly copied, external documentation relevant to solution
 -->
This PR creates a new `composite.json` file which contains composite tokens made up from our font family, line height, paragraph spacing, and letter spacing tokens.
- Add `filter/font/composite-npm` filter for font tokens of type `composite`
- Add `typescript/es6-declarations/composite` format to generate TS declarations composite tokens
  - Handles tokens with `value`s of object type
  - Infers value types using `typeof`
  - Includes comment for TSdoc/IDEs 
- The JS export for this is `font.styles`. I realize `styles` can be a little confusing alongside `style.json` which contains Bold, Italic, etc. but that file is for Figma only and not exported to NPM. Open to different names for this export.

Check [npm](https://www.npmjs.com/package/@department-of-veterans-affairs/mobile-tokens/v/0.19.1-alpha.4) for sample output

#### Testing Packages
<!-- List or range of alpha/beta packages published in association with this PR, if any -->
- 0.19.1-alpha.0
- 0.19.1-alpha.1
- 0.19.1-alpha.2
- 0.19.1-alpha.3
- 0.19.1-alpha.4


### Screenshots/Video
<!-- Add screenshots or video as needed; before/after recommended if appropriate. 
Convenience side-by-side formatting:
Before/after: <img src="" width="49%" />&nbsp;&nbsp;<img src="" width="49%" />
Accordion before/after: <details><summary>Before/after</summary><img src="" width="49%" />&nbsp;&nbsp;<img src="" width="49%" /></details>
-->
N/A

## Testing
<!-- Describe testing conducted to validate changes.
Consider highlighting:
- What testing was not explicitly done and may be relevant for QA? 
- Edge cases validated
- Special situations that could not be tested
- Any testing performed in a consuming app -->

Tested by importing into components package.

## PR Checklist
Code reviewer validation:
- General
	- [x] PR is linked to ticket(s)
	- [x] PR has `changelog` label applied if it's to be included in the changelog
	- [ ] Acceptance criteria: 
		- All satisfied _or_
		- Documented reason for not being performed _or_
		- Split to separate ticket and ticket is linked by relevant AC(s)
	- [x] Above PR sections adequately filled out
	- [x] If any breaking changes, in accordance with the [pre-1.0.0 versioning guidelines](https://github.com/department-of-veterans-affairs/va-mobile-library#versioning-policy): a CU ticket has been created for the VA Mobile App detailing necessary adjustments with the package version that will be published by this ticket
- Code
	- [x] Tests are included if appropriate (or split to separate ticket)
	- [x] New functions have proper TSDoc annotations

## Publish
<!-- Most changes entail a version increment; section can be removed for PRs exclusively within non-ship-relevant files (e.g. unit tests, Storybook stories) -->
If changes warrant a new version [per the versioning guidelines](https://github.com/department-of-veterans-affairs/va-mobile-library/blob/main/documentation/versioning.md) and the PR is approved and ready to merge:
- [ ] Merge `main` into branch
- [ ] Merge branch to `main`
- [ ] Verify that [Check Component Integrations](https://github.com/department-of-veterans-affairs/va-mobile-library/actions/workflows/check-component-integrations.yml) workflow ran successfully
- [ ] [Publish new version](https://github.com/department-of-veterans-affairs/va-mobile-library/actions/workflows/publish.yml) after merging

narin commented:
@jessicawoodin Could you please review [composite.json](https://github.com/department-of-veterans-affairs/va-mobile-library/pull/572/files#diff-00b448a1ce04a0c24ffe99a1d4b5160a1414e75a33d91946651564c1f23baa68) to check the values for accuracy? Thanks.

TimRoe commented on packages/tokens/src/tokens/font/composite.json:
```diff
@@ -0,0 +1,152 @@
+{
```
Are these not _the_ typography tokens? `composite` feels like very generic naming.

narin replied:
Renamed to typography

TimRoe commented on packages/tokens/config.js:
```diff
@@ -83,6 +83,14 @@ module.exports = {
+          format: 'javascript/es6/simple-key-value',
+          filter: 'filter/font/composite-npm',
+          options: {
+            exportName: 'styles',
```
```suggestion
            exportName: 'typography',
```
I'd always been under the impression using the tokens directly would be `typography.whatever` given they're the typography tokens.

TimRoe replied:
Also suggest updating the destination files accordingly.

TimRoe commented on packages/tokens/config.js:
```diff
@@ -83,6 +83,14 @@ module.exports = {
           format: 'javascript/es6/vads-colors-themes',
         },
+        {
+          destination: 'js/font/styles.js',
```
Should be prior to themes/spacing for alphabetical ordering based on destination.

TimRoe commented on packages/tokens/build.js:
```diff
@@ -249,6 +255,38 @@ StyleDictionary.registerFormat({
+
 /** Creates named type declaration for Themes. Allows for TypeScript autocomplete */
 StyleDictionary.registerFormat({
   name: 'typescript/es6-declarations/theme',
```
Just for confirmation, we aren't including Figma export of these tokens because:
- The tooling doesn't support composite tokens OR
- Figma can't meaningfully apply composite tokens similarly to atomic tokens OR
- Some other reason
?

I recall there being a reason, but don't recall what it was or if it'd been documented anywhere.

narin replied:
Are you referring to themes or typography?

narin replied:
Just confirmed with @jessicawoodin that Figma doesn't accept composite tokens so there's no need to export them

TimRoe commented on packages/tokens/build.js:
```diff
@@ -249,6 +255,38 @@ StyleDictionary.registerFormat({
+  name: 'typescript/es6-declarations/composite',
+  formatter: function ({ dictionary, options }) {
+    let tokens = dictionary.allTokens,
+      declaration = '`export declare const ${options.exportName}: {\n'
```
Looks like a last minute change broke things, getting the following error running `yarn tokens:build` against the branch:
```
dist/types/font/styles.d.ts
[error] dist/types/font/styles.d.ts: SyntaxError: Unterminated template literal. (82:2)
[error]   80 |     letterSpacing: number
[error]   81 |   }
[error] > 82 | }
[error]      |  ^
```

that looks related to the ` mark here turning the whole file into a string except the `${options.exportName}` bit.

TimRoe replied:
```suggestion
      declaration = `export declare const ${options.exportName}: {\n`
```

TimRoe commented on packages/tokens/config.js:
```diff
@@ -130,6 +138,15 @@ module.exports = {
+          filter: 'filter/font/composite-npm',
+          options: {
+            exportName: 'styles',
+            noSort: true,
```
Why is the typing not sorted while the JS is? Seems like they should be consistent. 

Given their composite nature not giving a clear numeric/size hierarchy, I think sorting both is fine?

TimRoe commented on packages/tokens/src/tokens/font/composite.json:
```diff
@@ -0,0 +1,152 @@
+{
+  "vads-font-display": {
```
Just adding a note that mostly deferring to Jessica for this file. General format looks fine, did not validate values.

TimRoe commented on packages/tokens/src/tokens/font/composite.json:
```diff
@@ -0,0 +1,152 @@
+      "fontFamily": "{vads-font-family-serif-regular}",
+      "fontSize": "{vads-font-size-display}",
+      "lineHeight": "{vads-font-line-height-heading-5xl}",
+      "marginBottom": "{vads-font-paragraph-spacing-heading-5xl}",
```
Weren't all these that were actively used handled by the spacing tokens? 

Thought that was the motivation for not including paragraph-spacing in the NPM dist; seems weird to reference Figma-only tokens within NPM tokens.

narin replied:
Chatted with @jessicawoodin about this over the past few days and decided we'll use spacing tokens instead. Composite tokens have been updated to use spacing tokens as marginBottom instead of paragraph spacing.

TimRoe reviewed (changes requested):
Three big things:
1. A late change broke the TS output file which needs fixing
2. The tacked-on AC about adding groupings to the Figma output didn't fall into this ticket as it didn't touch the Figma output, it should be propagated to a standalone ticket to ensure we don't forget that was still desired by design/Jessica
3. I think all instances of `styles` and `composite` should be `typography`

Searching typography gives the following from Wikipedia:
> Typography is the art and technique of arranging type to make written language legible, readable and appealing when displayed. The arrangement of type involves selecting typefaces, point sizes, line lengths, line spacing, letter spacing, and spaces between pairs of letters.

which is exactly what these tokens are: an amalgamation of those base font building blocks into a coherent unit. While we don't imminently have any future sets of composite tokens I've aware of, composite seems like a general classification of tokens made from other tokens while typography are the specific tokens that these composites are and these should be named accordingly.

Beyond that, just some other minor suggestions/comments/questions.

jessicawoodin reviewed (changes requested):
Overall, this PR looks great to me. All of the values appear correct.

Narin and I re-visited the paragraph spacing tokens based on how things were going with the Text component. We decided to remove all paragraph spacing tokens, and only use spacing tokens instead. I will update Figma and the Design Tokens spreadsheet to reflect this change. In the meantime, attached is an image showing the new spacing tokens that should be used for marginBottom in the composite tokens.

![image](https://github.com/user-attachments/assets/e376d83b-4f85-4080-a4be-b60563ba51cf)

narin commented on packages/tokens/build.js:
```diff
@@ -249,6 +255,38 @@ StyleDictionary.registerFormat({
+      })
+
+      docs += `*/\n`
+      declaration += `  ${docs}`
```
This seems to be resolved now that we're using the spacing tokens instead, which have integers instead of numbers with decimals

narin commented:
- Renamed composite/styles to typography
- Updated marginBottom (paragraph spacing) in composite tokens to use spacing tokens instead
- Fixed config files sorting

Ready for second review

TimRoe commented:
> Created separate ticket to look into combining figma export into single file: #577

Was there some further discussion I missed? If not, I think this ticket is a bit off. The font file for figma is already one file, it's just not grouped so things don't parse out into style/size/etc. in Figma as desired. This is the first I'm hearing of any desire to roll spacing/colors into a single file.

TimRoe commented on packages/tokens/src/tokens/font/typography.json:
```diff
@@ -0,0 +1,152 @@
+    },
+    "attributes": {
+      "category": "font",
+      "type": "composite",
```
Should this also be updated to `typography`?

TimRoe replied:
If so, think the filter name would also need to be updated.

TimRoe reviewed (approved):
Approved with just one last optional minor comment/suggestion.
